src/xargs/mod.rs

```diff
@@ -637,6 +640,38 @@ where
     }
 }
 
+struct EofArgumentReader {
+    reader: Box<dyn ArgumentReader>,
+    eof_delimiter: OsString,
+}
+
+impl EofArgumentReader {
+    fn new(reader: Box<dyn ArgumentReader>, eof_delimiter: &String) -> Self {
+        Self {
+            reader,
+            eof_delimiter: eof_delimiter.into(),
+        }
+    }
+}
+
+impl ArgumentReader for EofArgumentReader {
+    fn next(&mut self) -> io::Result<Option<Argument>> {
+        match self.reader.next() {
+            Err(e) => Err(e),
+            Ok(arg) => match arg {
+                None => Ok(None),
+                Some(arg) => {
+                    if arg.arg == self.eof_delimiter {
+                        Ok(None)
+                    } else {
+                        Ok(Some(arg))
+                    }
+                }
+            },
+        }
```

Comment:
I think you can simplify the code a bit by using `map`. Something like:
```suggestion
        self.reader.next().map(|opt| {
            opt.and_then(|arg| {
                if arg.arg == self.eof_delimiter {
                    None
                } else {
                    Some(arg)
                }
            })
        })
```